SimG4CMS/HGCalTestBeam/plugins/HGCPassive.cc

```diff
@@ -79,67 +80,84 @@ void HGCPassive::update(const BeginOfRun * run) {
-		  << aStep->GetTrack()->GetDefinition()->GetParticleName()
-		  << " at" << aStep->GetTrack()->GetPosition() << " Volume " 
-		  << aStep->GetTrack()->GetVolume() << ":" 
-		  << aStep->GetTrack()->GetNextVolume() << " Status " 
-		  << aStep->GetTrack()->GetTrackStatus() << " KE " 
-		  << aStep->GetTrack()->GetKineticEnergy() << " Deposit " 
-		  << aStep->GetTotalEnergyDeposit() << " Map "
-		  << (it != mapLV_.end()) << std::endl;
+	edm::LogVerbatim("ValidHGCal") << plv->GetName() << " F|L Step " 
+				       << aStep->IsFirstStepInVolume() << ":" 
+				       << aStep->IsLastStepInVolume() 
+				       << " Position" << aStep->GetPreStepPoint()->GetPosition()
+				       << " Track " << aStep->GetTrack()->GetDefinition()->GetParticleName()
+				       << " at" << aStep->GetTrack()->GetPosition() 
+				       << " Volume " << aStep->GetTrack()->GetVolume()
+				       << ":" << aStep->GetTrack()->GetNextVolume()
+				       << " Status " << aStep->GetTrack()->GetTrackStatus()
+				       << " KE " << aStep->GetTrack()->GetKineticEnergy()
+				       << " Deposit " << aStep->GetTotalEnergyDeposit()
+				       << " Map " << (it != mapLV_.end());
 #endif
-	double time   = aStep->GetTrack()->GetGlobalTime();
-	double energy = (aStep->GetPreStepPoint()->GetKineticEnergy() +
-			 aStep->GetTotalEnergyDeposit())/CLHEP::GeV;
-	if (it != mapLV_.end()) {
-	  storeInfo(it, plv, 0, time, energy);
-	} else if (topLV_ != nullptr) {
-	  auto itr = (init_) ? mapLV_.find(topLV_) : findLV(topLV_);
-	  if (itr != mapLV_.end()) {
-	    storeInfo(itr, topLV_, 0, time, energy);
-	  }
-	}
-      } else if (it != mapLV_.end()) {
-	unsigned int copy = (unsigned int)(touchable->GetReplicaNumber(0) + 
-					   1000*touchable->GetReplicaNumber(1));
-	double time = (aStep->GetPostStepPoint()->GetGlobalTime());
-	double edeposit = (aStep->GetTotalEnergyDeposit())/CLHEP::GeV;
-	storeInfo(it, plv, copy, time, edeposit);
+	energy += (aStep->GetPreStepPoint()->GetKineticEnergy()/CLHEP::GeV);
+      } else {
+	time = (aStep->GetPostStepPoint()->GetGlobalTime());
+	copy = (unsigned int)(touchable->GetReplicaNumber(0) + 
+			      1000*touchable->GetReplicaNumber(1));
+      }
+      if (it != mapLV_.end()) {
+	storeInfo(it, plv, copy, time, energy, true);
       } else if (topLV_ != nullptr) {
-	auto itr = findLV(topLV_);
+	auto itr = (init_) ? mapLV_.find(topLV_) : findLV(topLV_);
 	if (itr != mapLV_.end()) {
-	  double time = (aStep->GetPostStepPoint()->GetGlobalTime());
-	  double edeposit = (aStep->GetTotalEnergyDeposit())/CLHEP::GeV;
-	  storeInfo(itr, topLV_, 100, time, edeposit);
+	  storeInfo(itr, topLV_, copy, time, energy, true);
 	}
       }
     }//if (curSD==NULL)
+
+    //Now for the mother volumes
+    int level = (touchable->GetHistoryDepth());
+    if (level > 0) {
+      double energy = (aStep->GetTotalEnergyDeposit())/CLHEP::GeV;
+      double time   = (aStep->GetTrack()->GetGlobalTime());
+
+      for (int ii=0; ii<level; ++ii) {
```

Comment:
how about

for ( int i=level; i>0; i--) {

Reply:
Done